File: advertising_sdk/src/main/java/com/adadapted/android/sdk/core/ad/ImpressionIdCounter.java
package com.adadapted.android.sdk.core.ad;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class ImpressionIdCounter {
    private static ImpressionIdCounter instance;

    public static ImpressionIdCounter getsInstance() {
        if(instance == null) {
            instance = new ImpressionIdCounter();
        }

        return instance;
    }

    private static final int INITIAL_VAL = 1;

    private final Map<String, Integer> idCounts;
    private final Lock counterLock = new ReentrantLock();

    private ImpressionIdCounter() {
        idCounts = new HashMap<>();
    }

    private void initCountFor(final String impressionId) {
        idCounts.put(impressionId, INITIAL_VAL);
    }

    public synchronized int getIncrementedCountFor(final String impressionId) {
        counterLock.lock();
        try {
            if (idCounts.containsKey(impressionId)) {
                int val = idCounts.get(impressionId);
                val++;
                idCounts.put(impressionId, val);

                return val;
            } else {
                initCountFor(impressionId);
                return idCounts.get(impressionId);
            }
        }
        finally {
            counterLock.unlock();
        }
    }

    public synchronized int getCurrentCountFor(final String impressionId) {
        counterLock.lock();
        try {
            if (idCounts.containsKey(impressionId)) {
                return idCounts.get(impressionId);
            } else {
                initCountFor(impressionId);
                return idCounts.get(impressionId);
            }
        }
        finally {
            counterLock.unlock();
        }
    }
}

File: advertising_sdk/src/main/java/com/adadapted/android/sdk/core/ad/ImpressionIdCounter.kt
package com.adadapted.android.sdk.core.ad

import java.util.concurrent.locks.Lock
import java.util.concurrent.locks.ReentrantLock

class ImpressionIdCounter private constructor(): Counter {
    private val idCounts: MutableMap<String, Int?>
    private val counterLock: Lock = ReentrantLock()
    private fun initCountFor(impressionId: String) {
        idCounts[impressionId] = INITIAL_VAL
    }

    @Synchronized
    override fun getIncrementedCountFor(impressionId: String): Int {
        counterLock.lock()
        return try {
            if (idCounts.containsKey(impressionId)) {
                var value = idCounts[impressionId]!!
                value++
                idCounts[impressionId] = value
                value
            } else {
                initCountFor(impressionId)
                idCounts[impressionId]!!
            }
        } finally {
            counterLock.unlock()
        }
    }

    @Synchronized
    override fun getCurrentCountFor(impressionId: String): Int {
        counterLock.lock()
        return try {
            if (idCounts.containsKey(impressionId)) {
                idCounts[impressionId]!!
            } else {
                initCountFor(impressionId)
                idCounts[impressionId]!!
            }
        } finally {
            counterLock.unlock()
        }
    }

    companion object {
        var instance: ImpressionIdCounter? = null
            get() {
                if (field == null) {
                    field = ImpressionIdCounter()
                }
                return field
            }
            private set

        private const val INITIAL_VAL = 1
    }

    init {
        idCounts = HashMap()
    }
}